Sources/Basic/DiagnosticsEngine.swift

```diff
@@ -0,0 +1,269 @@
+        case literalItem(String, preference: Preference)
+
+        /// A substitution of a computed value.
+        case substitutionItem((DiagnosticData) -> DiagnosticParameter, preference: Preference)
+
+        public static func literal(
+            _ string: String,
+            preference: Preference = .default
+        ) -> DescriptionFragment {
+            return .literalItem(string, preference: preference)
+        }
+
+        public static func substitution(
+            _ accessor: @escaping ((DiagnosticData) -> DiagnosticParameter),
+            preference: Preference = .default
+        ) -> DescriptionFragment {
+            return .substitutionItem(accessor, preference: preference)
+        }
+    }
+    
+    /// The name of the diagnostic, which is expected to be in reverse dotted notation.
+    public let name: String
+
+    /// The English format string for the diagnostic description.
+    public let description: [DescriptionFragment]
+    
+    /// The default behavior associated with this diagnostic.
+    public let defaultBehavior: Diagnostic.Behavior
+
+    /// Create a new diagnostic identifier.
+    ///
+    /// - Parameters:
+    ///   - type: The type of the payload data, used to help type inference.
+    ///
+    ///   - name: The name of the identifier.
+    ///
+    ///   - description: A closure which will compute the description from a
+    ///                  builder. We compute descriptions in this fashion in
+    ///                  order to take advantage of type inference to make it
+    ///                  easy to have type safe accessors for the payload
+    ///                  properties.
+    ///
+    /// The intended use is to support a convenient inline syntax for defining
+    /// new diagnostics, for example:
+    ///     
+    ///     struct TooFewLives: DiagnosticData {
+    ///         static var id = DiagnosticID(
+    ///             type: TooFewLives.self,
+    ///             name: "org.swift.diags.too-few-lives",
+    ///             description: { $0 <<< "cannot create a cat with" <<< { $0.count } <<< "lives" }
+    ///         )
+    ///     
+    ///         let count: Int
+    ///     }
+    public init<T: DiagnosticData>(
+        type: T.Type,
+        name: String,
+        defaultBehavior: Diagnostic.Behavior = .error,
+        description buildDescription: (DiagnosticDescriptionBuilder<T>) -> ()
+    ) {
+        self.name = name
+        self.description = DiagnosticDescriptionBuilder<T>().build(buildDescription)
+        self.defaultBehavior = defaultBehavior
+    }
+}
+
+/// The payload of a diagnostic.
+public protocol DiagnosticData {
+    /// The identifier of the diagnostic this payload corresponds to.
+    static var id: DiagnosticID { get }
+}
+
+/// The location of the diagnostic.
+public protocol DiagnosticLocation {
+    /// The human readable summary description for the location.
+    var localizedDescription: String { get }
+}
+
+public struct Diagnostic {
```

Comment:
FYI, @abertelrud discussed a possible improvement here to do later where we would see if we can make merge DiagnosticData and Diagnostic so there is only one initializer. They are conceptually one object, it just happened it was easier to make the type system work out as written here. But I think it might be possible to merge so I put it on my TODO list. Ha Ha. :)